packages/flutter_tools/templates/app_shared/web/flutter.js.tmpl

```diff
@@ -0,0 +1,132 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/**
+ * This script installs service_worker.js to provide PWA functionality to
+ *     application. For more information, see:
+ *     https://developers.google.com/web/fundamentals/primers/service-workers
+ */
+
+if (!_flutter) {
+  var _flutter = {};
+}
+_flutter.loader = null;
+
+(function() {
+  "use strict";
+  class FlutterLoader {
+    // TODO: Move the below methods to "#private" once supported by all the browsers
+    // we support. In the meantime, we use the "revealing module" pattern.
+
+    // Watchdog to prevent injecting the main entrypoint multiple times.
+    _scriptLoaded = false;
+
+    // Resolver for the pending promise returned by loadMainDartJs.
+    _didLoadMainDartJsResolve = null;
+
+    /**
+     * Initializes the main.dart.js with/without serviceWorker.
+     * @param {*} options 
+     * @returns a Promise that will eventually resolve with an EngineInitializer,
+     * or will be rejected with the error caused by the loader.
+     */
+    loadMainDartJs(options) {
+      const { 
+        entrypointUrl = 'main.dart.js', 
+        serviceWorker,
+      } = options;
+      return this._loadWithServiceWorker(entrypointUrl, serviceWorker);
+    }
+
+    /**
+     * Resolves the promise created by loadMainDartJs. Called by Flutter.
+     * Needs to be weirdly bound like it is, so "this" is preserved across
+     * the JS <-> Flutter jumps.
+     * @param {*} engineInitializer 
+     */
+    didLoadMainDartJs = (function(engineInitializer) {
+      if (typeof this._didLoadMainDartJsResolve != 'function') {
+        console.warn('Do not call didLoadMainDartJs by hand. Start with loadMainDartJs instead.');
+      }
+      this._didLoadMainDartJsResolve(engineInitializer);
+    }).bind(this);
+
+    _loadMainDartJs(entrypointUrl) {
+      if (this._scriptLoaded) {
+        return null;
+      }
+  
+      this._scriptLoaded = true;
+  
+      return new Promise((resolve, reject) => {
+        let scriptTag = document.createElement('script');
+        scriptTag.src = entrypointUrl;
+        scriptTag.type = 'application/javascript';
+        this._didLoadMainDartJsResolve = resolve; // Cache the resolve, so it can be called from Flutter.
```

Comment:
When this Promise is resolved, all we have is an instance of `FlutterEngineInitializer` but the engine hasn't been initialized yet, right?

Reply:
Yes, only when you call engineInitializer.initEngine(), flutter does the initialization!